File: go/internal/der/der.go
// Package der is a bounded BER/DER reader — just enough ASN.1 to walk
// CMS/PKCS#7 SignedData and Apple receipt payloads.
//
// It is hand-rolled on purpose. encoding/asn1 is DER-only and refuses the
// indefinite lengths that genuine Apple and Xcode receipts use (10 of the
// 11 receipt fixtures in this repository are indefinite-length BER), and
// golang.org/x/crypto/cryptobyte rejects the 0x80 length octet for the
// same reason. PLAN.md D8 already settled the same question for the Node
// port: a small audited reader beats a large dependency in a security
// library.
//
// Every input this package sees is attacker-controlled. It therefore:
//
//   - bounds nesting depth (recursion is the obvious DoS),
//   - bounds the total node count (many tiny TLVs are the other one),
//   - rejects multi-byte tags and long-form lengths above four octets,
//   - checks every computed offset against the input length and against
//     integer overflow before slicing,
//   - rejects trailing bytes after the outermost value, and
//   - never panics: Parse returns an error for every malformed input.
package der

import (
	"errors"
	"fmt"
)

// MaxDepth is the deepest nesting Parse will follow. Real receipts nest
// about six levels; 32 leaves room to spare while making a nesting bomb a
// rejection rather than a stack overflow.
const MaxDepth = 32

// maxNodes bounds the total TLVs one Parse may produce.
//
// Every node costs at least two input bytes but around a hundred bytes of
// parsed tree, so a megabyte of two-byte TLVs is a fifty-megabyte
// allocation. Measured, the genuine 79 KB legacy receipt with 187 in-app
// purchases parses to 271 nodes and the sandbox one to 277, so this sits
// roughly 370x above anything real while capping the amplification at a
// constant.
const maxNodes = 100_000

// ErrTrailingBytes is returned when a complete value is followed by more
// bytes. Rejecting the remainder rather than ignoring it is what stops a
// verified CMS blob from carrying an unverified tail (PLAN.md §2.3).
var ErrTrailingBytes = errors.New("trailing bytes after ASN.1 value")

// Common identifier octets.
const (
	TagInteger                byte = 0x02
	TagBitString              byte = 0x03
	TagOctetString            byte = 0x04
	TagOID                    byte = 0x06
	TagUTF8String             byte = 0x0c
	TagIA5String              byte = 0x16
	TagSequence               byte = 0x30
	TagSet                    byte = 0x31
	TagOctetStringConstructed byte = 0x24
	TagContext0               byte = 0xa0
	TagContext1               byte = 0xa1
	TagContext2               byte = 0xa2
	TagContext3               byte = 0xa3
)

// Node is one parsed TLV.
//
// Raw and Contents are always sub-slices of the buffer handed to Parse,
// indefinite-length nodes included: the children of an indefinite-length
// value are contiguous between its length octet and its end-of-contents
// marker, so its contents need no concatenation. Callers that hand a byte
// field to their own caller must copy it — see the aliasing rule in the
// root package.
type Node struct {
	Tag         byte
	Constructed bool
	Raw         []byte
	Contents    []byte
	Children    []*Node
}

type parser struct {
	buf   []byte
	nodes int
}

// Parse reads exactly one ASN.1 value from b and rejects anything left
// over.
func Parse(b []byte) (*Node, error) {
	p := &parser{buf: b}
	node, end, err := p.readNode(0, 0)
	if err != nil {
		return nil, err
	}
	if end != len(b) {
		return nil, fmt.Errorf("%w (%d)", ErrTrailingBytes, len(b)-end)
	}
	return node, nil
}

func (p *parser) readNode(off, depth int) (*Node, int, error) {
	if depth > MaxDepth {
		return nil, 0, fmt.Errorf("maximum ASN.1 nesting depth (%d) exceeded", MaxDepth)
	}
	p.nodes++
	if p.nodes > maxNodes {
		return nil, 0, fmt.Errorf("ASN.1 value has more than %d nodes", maxNodes)
	}
	if off < 0 || off+2 > len(p.buf) {
		return nil, 0, errors.New("truncated ASN.1 value")
	}
	tag := p.buf[off]
	if tag&0x1f == 0x1f {
		return nil, 0, errors.New("multi-byte ASN.1 tags are not supported")
	}
	constructed := tag&0x20 != 0
	pos := off + 1
	lenByte := p.buf[pos]
	pos++

	indefinite := false
	length := 0
	switch {
	case lenByte < 0x80:
		length = int(lenByte)
	case lenByte == 0x80:
		if !constructed {
			return nil, 0, errors.New("indefinite length on a primitive value")
		}
		indefinite = true
	default:
		n := int(lenByte & 0x7f)
		// Four octets caps a declared length at 2^32-1, which is far above
		// any receipt and keeps the accumulation below int64 overflow. The
		// bounds check happens before any length octet is read.
		if n > 4 || pos+n > len(p.buf) {
			return nil, 0, errors.New("unsupported ASN.1 length")
		}
		for i := 0; i < n; i++ {
			length = length<<8 | int(p.buf[pos+i])
		}
		pos += n
	}

	if !indefinite {
		end := pos + length
		// end < pos catches the overflow case on any word size; end >
		// len(buf) catches a length that claims more than we have.
		if end < pos || end > len(p.buf) {
			return nil, 0, errors.New("ASN.1 length exceeds input")
		}
		node := &Node{
			Tag:         tag,
			Constructed: constructed,
			Raw:         p.buf[off:end],
			Contents:    p.buf[pos:end],
		}
		if constructed {
			children, err := p.readChildren(pos, end, depth+1)
			if err != nil {
				return nil, 0, err
			}
			node.Children = children
		}
		return node, end, nil
	}

	// Indefinite (BER) length: children until an end-of-contents marker.
	//
	// The children run contiguously from here to the marker, so the
	// contents are the slice between the two — NOT a concatenation of the
	// children's Raw bytes. Building that concatenation would copy the
	// whole subtree once per level of nesting, which is O(MaxDepth x
	// input) and is capped by neither maxNodes (a node count) nor
	// MaxDepth (a multiplier on the copy, not a bound on it): 30 levels
	// of indefinite wrapping cost four bytes of input each and multiplied
	// the parse by 30.
	contentStart := pos
	var children []*Node
	for {
		if pos+2 > len(p.buf) {
			return nil, 0, errors.New("unterminated indefinite-length value")
		}
		if p.buf[pos] == 0x00 && p.buf[pos+1] == 0x00 {
			pos += 2
			break
		}
		child, next, err := p.readNode(pos, depth+1)
		if err != nil {
			return nil, 0, err
		}
		children = append(children, child)
		pos = next
	}
	return &Node{
		Tag:         tag,
		Constructed: true,
		Raw:         p.buf[off:pos],
		Contents:    p.buf[contentStart : pos-2],
		Children:    children,
	}, pos, nil
}

func (p *parser) readChildren(start, end, depth int) ([]*Node, error) {
	var children []*Node
	pos := start
	for pos < end {
		child, next, err := p.readNode(pos, depth)
		if err != nil {
			return nil, err
		}
		if next > end {
			return nil, errors.New("ASN.1 child overruns its parent")
		}
		children = append(children, child)
		pos = next
	}
	return children, nil
}

// IsOctetString reports whether n is an OCTET STRING in either the
// primitive or the BER constructed form.
func IsOctetString(n *Node) bool {
	return n != nil && (n.Tag == TagOctetString || n.Tag == TagOctetStringConstructed)
}

// OctetValue is the value of an OCTET STRING, joining the chunks of a BER
// constructed one. Depth is already bounded by Parse.
//
// The flattening is a single pass into one buffer sized in advance. The
// obvious recursive spelling — concatenate each child's OctetValue —
// re-copies the whole subtree at every level on the way up, so a
// 28-deep constructed OCTET STRING costs 28x its own size to read. That
// is attacker-controlled nesting and it is not bounded by the node cap.
func OctetValue(n *Node) []byte {
	if n == nil {
		return nil
	}
	if !n.Constructed {
		return n.Contents
	}
	out := make([]byte, 0, octetValueLen(n))
	return appendOctetValue(out, n)
}

// octetValueLen is the exact size of OctetValue's answer, so the buffer
// is allocated once and never grown.
func octetValueLen(n *Node) int {
	if !n.Constructed {
		return len(n.Contents)
	}
	total := 0
	for _, child := range n.Children {
		total += octetValueLen(child)
	}
	return total
}

func appendOctetValue(dst []byte, n *Node) []byte {
	if !n.Constructed {
		return append(dst, n.Contents...)
	}
	for _, child := range n.Children {
		dst = appendOctetValue(dst, child)
	}
	return dst
}

// Child returns the i-th child, or nil when there is none. Returning nil
// rather than panicking is what lets the CMS walk read a structure of an
// unexpected shape without an index check at every step.
func Child(n *Node, i int) *Node {
	if n == nil || i < 0 || i >= len(n.Children) {
		return nil
	}
	return n.Children[i]
}

File: php/src/Internal/Der.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Internal;

/**
 * Minimal DER/BER reader — just enough ASN.1 to walk X.509 certificates,
 * CMS/PKCS#7 SignedData and Apple receipt payloads. Hand-rolled on purpose
 * (PLAN.md §1 + D8): dependency-light and auditable, and a port of the same
 * reader the Node implementation uses, so the two cannot disagree about what
 * a receipt says.
 *
 * Definite and indefinite (BER 0x80) lengths are both supported because
 * genuine Apple and Xcode receipts use both.
 *
 * Everything this class reads is attacker-supplied, so it is bounded in four
 * independent ways: nesting depth, node count, retained bytes, and the length
 * encodings it will honour. Those budgets are PHP-specific requirements rather
 * than ports of Node constants: PHP has no zero-copy slice, so every value is
 * a fresh string and every node a real object. Measured on PHP 8.4, a 1 MB
 * input of minimal two-byte nodes costs about 72 MB of parser state — a 76×
 * amplification, against a `php.ini-production` default `memory_limit` of
 * 128M. The Node port has no such failure mode because its subarrays are
 * views over one buffer.
 *
 * ## Why depth and node count together are still not enough
 *
 * Those two bound different axes, and the cost is their PRODUCT: a value
 * nested N levels deep is copied N times on the way down, so retained state is
 * roughly `2 × depth × input`. Deep-but-large nesting is cheap on every axis
 * they bound — 600 sibling chains of 31 SEQUENCEs around 3 KB each is 19,201
 * nodes at depth 31 in 1.9 MB of input, inside the node budget, the depth
 * ceiling and the receipt byte cap alike — and cost 92 MB of parser state,
 * which is a `memory_limit` exhaustion on a default deployment. That is a
 * fatal error and NOT a `Throwable`, so no `catch` in this library can turn it
 * into a verdict: the worker dies with no answer at all.
 * {@see DEFAULT_BYTE_BUDGET} is the bound on the product, and it is the one
 * that has to hold.
 *
 * @internal
 */
final class Der
{
    /** Nesting depth ceiling. Also what keeps recursion off PHP 8.1's unguarded stack. */
    public const MAX_DEPTH = 32;

    /**
     * Default ceiling on the number of nodes one parse may produce.
     *
     * The largest genuine fixture — the 79 KB legacy receipt with 187 in-app
     * purchases — decodes to well under 3,000 nodes in total, so this leaves
     * an order of magnitude of headroom over any real receipt.
     */
    public const DEFAULT_NODE_BUDGET = 20000;

    /**
     * Default ceiling on the bytes one parse may RETAIN — the sum, over every
     * node, of its `raw` plus its `contents`, which is what the tree actually
     * holds once `substr()` has copied both.
     *
     * The largest genuine fixture — the 79 KB legacy receipt with 187 in-app
     * purchases — retains 967 KB, a ratio of 12× that comes from real receipts
     * being shallow. The attack ratio is `2 × MAX_DEPTH`, i.e. 64×. So 32 MiB
     * is ~34× the largest real receipt while still covering a legitimate
     * receipt at the full {@see \EminDeniz99\ApplePurchaseReceiptVerifier\Receipt\ReceiptVerifier::DEFAULT_MAX_RECEIPT_BYTES}
     * at realistic ASN.1 shapes, and it caps hostile input at roughly a third
     * of a default 128M `memory_limit` instead of blowing straight through it.
     */
    public const DEFAULT_BYTE_BUDGET = 33554432;

    public const TAG_BOOLEAN = 0x01;
    public const TAG_INTEGER = 0x02;
    public const TAG_BIT_STRING = 0x03;
    public const TAG_OCTET_STRING = 0x04;
    public const TAG_OID = 0x06;
    public const TAG_UTF8_STRING = 0x0c;
    public const TAG_IA5_STRING = 0x16;
    public const TAG_UTC_TIME = 0x17;
    public const TAG_GENERALIZED_TIME = 0x18;
    public const TAG_OCTET_STRING_CONSTRUCTED = 0x24;
    public const TAG_SEQUENCE = 0x30;
    public const TAG_SET = 0x31;
    public const TAG_CONTEXT_0 = 0xa0;
    public const TAG_CONTEXT_1 = 0xa1;
    public const TAG_CONTEXT_2 = 0xa2;
    public const TAG_CONTEXT_3 = 0xa3;

    /**
     * Parses exactly one top-level value. Trailing bytes after it are an
     * error, not something to ignore: a CMS blob with garbage appended is a
     * different blob, and accepting it would let two ports disagree.
     *
     * @throws ParseException
     */
    public static function parse(
        string $buf,
        int $nodeBudget = self::DEFAULT_NODE_BUDGET,
        int $byteBudget = self::DEFAULT_BYTE_BUDGET,
    ): Asn1Node {
        $remaining = $nodeBudget;
        $bytes = $byteBudget;
        [$node, $end] = self::readNode($buf, 0, 0, $remaining, $bytes);
        if ($end !== strlen($buf)) {
            throw new ParseException('trailing bytes after ASN.1 value (' . (strlen($buf) - $end) . ')');
        }

        return $node;
    }

    /**
     * @param int $budget     node budget, consumed by reference across the whole tree
     * @param int $byteBudget retained-byte budget, likewise
     *
     * @return array{Asn1Node, int}
     *
     * @throws ParseException
     */
    private static function readNode(string $buf, int $off, int $depth, int &$budget, int &$byteBudget): array
    {
        if ($depth > self::MAX_DEPTH) {
            throw new ParseException('maximum ASN.1 nesting depth exceeded');
        }
        if ($budget <= 0) {
            throw new ParseException('ASN.1 node budget exceeded');
        }
        --$budget;

        $len = strlen($buf);
        if ($off + 2 > $len) {
            throw new ParseException('truncated ASN.1 value');
        }
        $tag = ord($buf[$off]);
        if (($tag & 0x1f) === 0x1f) {
            throw new ParseException('multi-byte ASN.1 tags are not supported');
        }
        $constructed = ($tag & 0x20) !== 0;
        $pos = $off + 1;
        $lenByte = ord($buf[$pos]);
        ++$pos;

        $length = null;
        if ($lenByte < 0x80) {
            $length = $lenByte;
        } elseif ($lenByte === 0x80) {
            if (!$constructed) {
                throw new ParseException('indefinite length on a primitive value');
            }
        } else {
            $numBytes = $lenByte & 0x7f;
            if ($numBytes > 4 || $pos + $numBytes > $len) {
                throw new ParseException('unsupported ASN.1 length');
            }
            $length = 0;
            for ($i = 0; $i < $numBytes; ++$i) {
                $length = $length * 256 + ord($buf[$pos + $i]);
            }
            $pos += $numBytes;
        }

        if ($length !== null) {
            $end = $pos + $length;
            if ($end > $len) {
                throw new ParseException('ASN.1 length exceeds input');
            }
            $contents = substr($buf, $pos, $length);
            $raw = substr($buf, $off, $end - $off);
            // Charged BEFORE descending, so a chain of nested wrappers around a
            // large value stops on the way down rather than after every level
            // has already copied it.
            self::charge($byteBudget, strlen($contents) + strlen($raw));
            $children = $constructed
                ? self::readChildren($contents, $depth + 1, $budget, $byteBudget)
                : null;

            return [new Asn1Node($tag, $constructed, $raw, $contents, $children), $end];
        }

        // Indefinite length: children until an end-of-contents (00 00) marker.
        $children = [];
        for (;;) {
            if ($pos + 2 > $len) {
                throw new ParseException('unterminated indefinite-length value');
            }
            if ($buf[$pos] === "\x00" && $buf[$pos + 1] === "\x00") {
                $pos += 2;
                break;
            }
            [$child, $next] = self::readNode($buf, $pos, $depth + 1, $budget, $byteBudget);
            $children[] = $child;
            $pos = $next;
        }
        $joined = '';
        foreach ($children as $child) {
            $joined .= $child->raw;
        }
        $raw = substr($buf, $off, $pos - $off);
        self::charge($byteBudget, strlen($joined) + strlen($raw));

        return [new Asn1Node($tag, true, $raw, $joined, $children), $pos];
    }

    /**
     * @param int $byteBudget consumed by reference
     *
     * @throws ParseException
     */
    private static function charge(int &$byteBudget, int $bytes): void
    {
        $byteBudget -= $bytes;
        if ($byteBudget < 0) {
            throw new ParseException('ASN.1 retained-byte budget exceeded');
        }
    }

    /**
     * @return list<Asn1Node>
     *
     * @throws ParseException
     */
    private static function readChildren(string $contents, int $depth, int &$budget, int &$byteBudget): array
    {
        $children = [];
        $pos = 0;
        $len = strlen($contents);
        while ($pos < $len) {
            [$child, $next] = self::readNode($contents, $pos, $depth, $budget, $byteBudget);
            $children[] = $child;
            $pos = $next;
        }

        return $children;
    }

    public static function isOctetString(Asn1Node $node): bool
    {
        return $node->tag === self::TAG_OCTET_STRING
            || $node->tag === self::TAG_OCTET_STRING_CONSTRUCTED;
    }

    /** Value bytes of an OCTET STRING, joining BER constructed chunks. */
    public static function octets(Asn1Node $node): string
    {
        if (!$node->constructed) {
            return $node->contents;
        }
        $out = '';
        foreach ($node->children() as $child) {
            $out .= self::octets($child);
        }

        return $out;
    }

    /**
     * DER-encodes an OBJECT IDENTIFIER dotted string to its contents bytes.
     * Marker-OID checks compare these bytes, never a name a crypto library
     * decided to give the OID.
     */
    public static function encodeOidContents(string $oid): string
    {
        $parts = array_map(intval(...), explode('.', $oid));
        if (count($parts) < 2) {
            throw new ParseException('not an OID: ' . $oid);
        }
        $out = chr(40 * $parts[0] + $parts[1]);
        for ($i = 2, $n = count($parts); $i < $n; ++$i) {
            $value = $parts[$i];
            $chunk = chr($value & 0x7f);
            $value = intdiv($value, 128);
            while ($value > 0) {
                $chunk = chr(($value & 0x7f) | 0x80) . $chunk;
                $value = intdiv($value, 128);
            }
            $out .= $chunk;
        }

        return $out;
    }

    /** Decodes OBJECT IDENTIFIER contents bytes back to a dotted string. */
    public static function decodeOid(string $contents): string
    {
        if ($contents === '') {
            throw new ParseException('empty OBJECT IDENTIFIER');
        }
        $first = ord($contents[0]);
        $parts = [min(intdiv($first, 40), 2)];
        $parts[] = $first - $parts[0] * 40;
        $value = 0;
        $started = false;
        for ($i = 1, $n = strlen($contents); $i < $n; ++$i) {
            $byte = ord($contents[$i]);
            $value = $value * 128 + ($byte & 0x7f);
            $started = true;
            if (($byte & 0x80) === 0) {
                $parts[] = $value;
                $value = 0;
                $started = false;
            }
        }
        if ($started) {
            throw new ParseException('truncated OBJECT IDENTIFIER');
        }

        return implode('.', $parts);
    }
}
